### DISCUSSION TITLE
Compare collections using a delegate

### DISCUSSION THREAD

**Discussion by vc-74** [xunit/xunit#2683] [Ideas]
`Assert` currently supports comparing collections, optionally passing an `IEqualityComparer` implementation.

If I'm not mistaken, only `IEqualityComparer.Equals` required in this case (IMO `IEqualityComparer` should have been called `IHashable` and inherit from `IEqualityComparer` which would have only `Equals` but I digress...).

The problem with that approach is that we have to build an `IEqualityComparer` even for simple cases, it would be much simpler to allow passing an inline comparer like:

```csharp
Assert.Equal(col1, col2, (expected, actual) =>
{
    // Comparison logic here
});
```

**Comment by bradwilson** [comment]
Thanks, I've converted this to #2692.

**Comment by vc-74** [comment]
Thanks @bradwilson!

**Comment by vc-74** [comment]
@bradwilson 
Sorry for reopening this discussion...

`Assert.Equal<T>(IEnumerable<T>?, IEnumerable<T>?, Func<T, T, bool>)` is a nice addition but I expected more something like:
`Assert.Equal<T>(IEnumerable<T>?, IEnumerable<T>?, Action<T, T>)` which would allow using assertions in the specified delegate and provide a better description in case of failure.

**Reply by bradwilson** [reply]
If you choose to throw an exception during the comparison, that exception will be reported (and assertions are exceptions). So you could write this:

![image](https://github.com/user-attachments/assets/0c738ee8-610b-40c3-a786-7e88363f6f23)

This would allow you to use whatever assertion and/or exception is appropriate for your situation, while maintaining the value provided by our invocation wrapper (showing the collections and pointers to the place where the issue occurred).

**Reply by vc-74** [reply]
OK, thanks
